src/gc.rs

```diff
@@ -0,0 +1,229 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use crate::{
+    alloc::{AllocMetadata, PtrInfo},
+    Gc,
+};
+use std::sync::Mutex;
+
+static WORD_SIZE: usize = 8; // Bytes
+
+type StackAddress = usize;
+type StackValue = usize;
+
+type StackScanCallback = extern "sysv64" fn(&mut Collector, StackAddress);
+#[link(name = "SpillRegisters", kind = "static")]
+extern "sysv64" {
+    // Rust (correctly) warns us that `Collector` has an unspecified layout and
+    // that it is UB to pass this to C code via FFI. However, our register spill
+    // code does not dereferences this, nor does it ever move the `Collector`
+    // instance by value. It simply moves the address to the first argument
+    // register to be used as the implicit `self` address in the callback. It is
+    // therefore safe to do this without specifying a C style layout for the
+    // struct.
+    #[allow(improper_ctypes)]
+    fn spill_registers(collector: &mut Collector, callback: StackScanCallback);
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub(crate) enum CollectorState {
+    Ready,
+    Marking,
+    FinishedMarking,
+}
+
+pub struct DebugFlags {
+    mark_only: bool,
+}
+
+impl DebugFlags {
+    pub fn new() -> Self {
+        Self { mark_only: false }
+    }
+
+    pub fn none() -> Self {
+        Self::new()
+    }
+
+    /// Prevents the GC from reclaiming 'garbage' objects. This flag is useful
+    /// for debugging whether the collector was able to find an object(s) while
+    /// traversing the object graph.
+    pub fn mark_only(mut self) -> Self {
+        self.mark_only = true;
+        self
+    }
+}
+
+/// Colour of an object used during marking phase (see Dijkstra tri-colour
+/// abstraction)
+#[derive(PartialEq, Eq)]
+pub(crate) enum Colour {
+    Black,
+    White,
+}
+
+pub(crate) struct Collector {
+    worklist: Vec<PtrInfo>,
+    black: bool,
+    debug_flags: DebugFlags,
+    pub(crate) state: Mutex<CollectorState>,
+}
+
+impl Collector {
+    pub(crate) fn new(debug_flags: DebugFlags) -> Self {
+        Self {
+            worklist: Vec::new(),
+            black: true,
+            debug_flags,
+            state: Mutex::new(CollectorState::Ready),
+        }
+    }
+
+    pub(crate) fn current_black(&self) -> bool {
+        self.black
+    }
+
+    pub(crate) fn collect(&mut self) {
+        // First check that no call to collect is active
+        {
+            let mut cstate = self.state.lock().unwrap();
+            match *cstate {
+                CollectorState::Ready => *cstate = CollectorState::Marking,
+                _ => {
+                    // Bail out if we reach this point as the collector has already
+                    // been invoked -- and is running -- on another thread.
```

Comment:
Simply saying "The collector is running in another thread" is probably good enough here?

Reply:
Fixed

Reply:
Laurie an I agreed to exhaustively enumerate (without `_`) in cases where it is practical, as it forces us to consider behaviours in the case that we add another variant.

Reply:
That makes sense,  but in this instance, collection can never happen when the collector is in any state other than `CollectorState::Ready`, even if new variants are added.

Reply:
Yeah, in this situation `_` did feel like it was reasonable to me (I explicitly thought about it!).

Reply:
It's up to you two. Seemed like a good candidate for the rule to me (few variants, not inconceivable that a new GC state be added later).